### PR TITLE
Feature/561/file sharing controller

### PR DESCRIPTION
**Pull request by lkleisa**
Controller for encryptable file sharing

### CODE REVIEW

File: app/models/encryptable.rb [hunk 1]
```diff
@@ -35,6 +38,10 @@ def decrypt(_team_password)
     raise 'implement in subclass'
   end
 
+  def decrypt_transfered(_private_key)
+    raise 'implement in subclass'
```
Comment (Robin481) [inline app/models/encryptable.rb:42]: huh?

File: app/models/encryptable.rb [hunk 2]
```diff
@@ -47,6 +54,21 @@ def team
     folder.team
   end
 
+  def recrypt_transferred(private_key, team_password)
+    decrypt(plaintext_transfer_password(private_key))
+    self.encrypted_transfer_password = nil
```
Comment (Robin481) [inline app/models/encryptable.rb:59]: why?

File: app/models/encryptable/file.rb [hunk 3]
```diff
@@ -7,9 +7,7 @@ class Encryptable::File < Encryptable
              class_name: 'Encryptable::Credentials',
              foreign_key: :credential_id
 
-  validates :credential_id, presence: true
```
Comment (Robin481) [inline app/models/encryptable/file.rb:10]: Why where these removed?

File: app/utils/encryptable_transfer.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  end
+
+  def receive(encryptable, private_key, personal_team_password)
+    encryptable.decrypt_transfered(private_key)
+
+    encryptable.update!(encrypted_transfer_password: nil,
+                        receiver_id: nil,
+                        sender_id: nil)
+
+    encryptable.encrypt(personal_team_password)
```
Comment (Robin481) [inline app/utils/encryptable_transfer.rb:24]: why encrypt and then instantly decrypt?

File: frontend/translations/de.yml [hunk 5]
```diff
@@ -41,9 +41,10 @@ de:
   new: Neu
   back: Zurück
   save: Speichern
+  share: Teilen
   close: Schliessen
   user: Benutzer
-  hi_user: "Hi, "
+  hi_user: "Moin, "
```
Comment (Robin481) [inline frontend/translations/de.yml:47]: Nein

File: spec/controllers/api/encryptables_controller_spec.rb [hunk 6]
```diff
@@ -696,7 +670,92 @@
...
+      received_file = Encryptable.find(shared_file.id)
+      file_content = fixture_file_upload('test_file.txt', 'text/plain').read
+
+      expect(received_file.encrypted_transfer_password).to eq(nil)
+      expect(received_file.sender_id).to eq(nil)
+      expect(received_file.name).to eq('test_file.txt')
+      expect(received_file.description).to eq('test')
+      expect(received_file.content_type).to eq('text/plain')
+      expect(received_file.folder_id).to eq(alice.inbox_folder.id)
+      expect(response.body).to eq file_content
```
Comment (Robin481) [inline spec/controllers/api/encryptables_controller_spec.rb:758]: Nice!

File: spec/unit/utils/encryptable_move_handler_spec.rb [hunk 7]
```diff
@@ -26,7 +26,7 @@
   end
 
   it 'moves credential from team2 folder to team1 folder' do
-    credential = encryptables(:credentials2)
+    credential = encryptables(:credentials3)
```
Comment (Robin481) [inline spec/unit/utils/encryptable_move_handler_spec.rb:29]: Is this because you moved the fixtures around?

**Review by Robin481 — CHANGES_REQUESTED** [review]

File: app/models/encryptable.rb [hunk 8]
```diff
@@ -71,4 +93,9 @@ def decrypt_attr(attr, team_password)
     instance_variable_set("@cleartext_#{attr}", cleartext_value)
   end
 
+  def assert_human_receiver?
```
[the remark this answers is not in the document]
Comment (lkleisa) [reply app/models/encryptable.rb:96]: Yes, but this comes from a validates

File: spec/controllers/api/encryptables_controller_spec.rb [hunk 9]
```diff
@@ -78,32 +80,10 @@
...
-
-      expect_json_object_includes_keys(credentials2_json_attributes, attributes)
-      expect_json_object_includes_keys(credentials2_json_relationships, nested_models)
-    end
-
     it 'returns encryptable files for credentials entry' do
       login_as(:alice)
 
       credentials1 = encryptables(:credentials1)
-      file1 = encryptables(:file1)
```
[the remark this answers is not in the document]
Comment (lkleisa) [reply spec/controllers/api/encryptables_controller_spec.rb:106]: Yes, we added encryptables(:file1) to global variables, because of that this line is not necessary

File: spec/unit/utils/encryptable_move_handler_spec.rb (hunk 7 above)
Comment (lkleisa) [reply spec/unit/utils/encryptable_move_handler_spec.rb:29]: We changed it, because a shared_file is attached to credentials2. And because we need to hardcode the encrypted data in fixtures, the MoveHandler couldn't decrypt this file. I created a new credential with a new attached file, this is working and tests the method

File: app/utils/encryptable_transfer.rb (hunk 4 above)
Comment (lkleisa) [reply app/utils/encryptable_transfer.rb:24]: The encryptable is decrypted with the private_key of the receiver before this line. After receive, we want to decrypt the encryptable with the personal_team_password, same as all other encryptables. Because of that, we encrypt it first with the personal_team_password. But the #receive function is called from the #show function of the encryptablesController. The #show function decrypts normal encryptables, because of that, we need to return a decrypted encryptable.

File: app/controllers/api/encryptables_controller.rb [hunk 10]
```diff
@@ -18,69 +17,105 @@ def index
...
+  def receiver_id
+    params.dig('data', 'attributes', 'receiver_id')
+  end
+
   # rubocop:disable Metrics/MethodLength
   def model_class
-    if create_ose_secret?
+    case action_name
+    when 'create'
+      define_model_class
```
[the remark this answers is not in the document]
Comment (lkleisa) [reply app/controllers/api/encryptables_controller.rb:73]: https://github.com/puzzle/cryptopus/issues/655

File: spec/controllers/api/encryptables_controller_spec.rb [hunk 11]
```diff
@@ -696,7 +670,92 @@
...
+        description: 'test',
+        data: {
+          attributes: {
+            receiver_id: api_user.id
+          }
+        }
+      }
+
+      expect { post :create, params: file_params, xhr: true }.
+        to raise_error(StandardError, 'Cant transfer to Api user')
```
Comment (mtnstar) [inline spec/controllers/api/encryptables_controller_spec.rb:695]: expect access denied instead of exception

File: spec/controllers/api/encryptables_controller_spec.rb [hunk 12]
```diff
@@ -696,7 +670,92 @@
...
+        description: 'test',
+        data: {
+          attributes: {
+            receiver_id: nil
+          }
+        }
+      }
+
+      expect { post :create, params: file_params, xhr: true }.
+        to raise_error(StandardError, 'Receiver user not found')
```
Comment (mtnstar) [inline spec/controllers/api/encryptables_controller_spec.rb:715]: expect record not found error, no custom error required

File: spec/controllers/api/encryptables_controller_spec.rb [hunk 13]
```diff
@@ -696,7 +670,92 @@
...
+    end
+
+    it 'transfers new file from sender to recipient' do
```
Comment (mtnstar) [inline spec/controllers/api/encryptables_controller_spec.rb:718]: how about
'bob sends file to alice' ?

File: spec/fixtures/encryptables.yml [hunk 14]
```diff
@@ -30,9 +30,24 @@ credentials2:
   name: Twitter Account
   description: My personal twitter account
   folder: folder2
-  tag: tag
   encrypted_data: '{"password":{"iv":null,"data":"X2i8woXXwIHew6zcnBws9Q=="},"username":{"iv":null,"data":"Kvkd66uUiNq4Gw4Yh7PvVg=="}}'
 
+credentials3:
```
Comment (mtnstar) [inline spec/fixtures/encryptables.yml:35]: do not add any more fixtures !

use fabricator to create entries where required !

File: spec/fixtures/files/users/api_tux/public.key [hunk 15]
```diff
@@ -0,0 +1,9 @@
+-----BEGIN PUBLIC KEY-----
```
Comment (mtnstar) [inline spec/fixtures/files/users/api_tux/public.key:1]: why is this required?

File: spec/unit/utils/encryptable_move_handler_spec.rb [hunk 16]
```diff
@@ -26,10 +26,19 @@
...
```
Comment (mtnstar) [inline spec/unit/utils/encryptable_move_handler_spec.rb:29]: use Fabricator or use existing fixture

File: spec/unit/utils/encryptable_transfer_spec.rb [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+      encryptable_transfer.transfer(encryptable_file, receiver, sender)
+
+      received_file = bob.personal_team.folders.find_by(name: 'inbox').encryptables.first
+      expect(received_file.encrypted_transfer_password).to be_present
+
+      personal_team_password = bob.personal_team.decrypt_team_password(bob, bobs_private_key)
+
+      encryptable_transfer.receive(received_file, bobs_private_key, personal_team_password)
+
+      expect(received_file.encrypted_transfer_password).to be_nil
```
Comment (mtnstar) [inline spec/unit/utils/encryptable_transfer_spec.rb:33]: maybe check received_file.transferred? false ?

**Review by mtnstar — CHANGES_REQUESTED** [review]
looks better and better.
still some things to check.

try to keep code as simple as possible. use dry crud defaults wherever possible. 
do not create custom raises unless really required.

File: spec/unit/utils/encryptable_transfer_spec.rb (hunk 17 above)
Comment (lkleisa) [reply spec/unit/utils/encryptable_transfer_spec.rb:33]: It's not possible, because file.transferred? checks for encrypted_transfer_password and in this state, the transfer_password is already set to nil
